---
title: "CACIMAR tutorial"
author:  Jiang junyao
date: "`r Sys.Date()`"
output:
  rmdformats::downcute:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    highlight: tango
---

<style>
  body {
  font-family: "Helvetica";
    font-size: 18px;
    color: black;
  }
</style>

```{r setup, include=FALSE}
## Global options
knitr::opts_chunk$set(cache = TRUE)
```

# Installation

Follow the steps below to install CACIMAR package from GitHub and run it:

``` r
# install
# install.packages("devtools") # Install devtools if not already installed
devtools::install_github("jiang-junyao/CACIMAR")
# require
library(CACIMAR)
library(reshape2)
```

# 1.Identify markers
  In this part, CACIMAR first uses receiver operating characteristic (ROC) based method in 'FindAllMarkers' function of [Seurat](https://satijalab.org/seurat/articles/get_started.html) to identify markers in each cluster for every species individually. Then, based on marker genes identified above, CACIMAR calculates the power of markers(MP, MP= 2 × abs(AUC-0.5)) in each cluster and the expression differences of markers(logFC) between clusters. Markers with MP > 0.35 and logFC > 0.1 will be retained. 
  
## (1) Inputs data of this part

- ***Seurat object***
  Seurat object should have cell types information in 'active.ident' slot. Test data used here can be download from [Google drive](https://drive.google.com/drive/folders/15bqQmEH4cDFRiEHYZTEp4gYWK4FJAhBL?usp=sharing)

### (2) Identify markers

```r
### identify markers
Zf_seurat <- readRDS('Zf_seurat.rds')
Mm_seurat <- readRDS('Mm_seurat.rds')
Zf_marker <- Identify_Markers(Zf_seurat,PowerCutoff=0.35, DifferenceCutoff=0.1,PvalueCutoff=0.05)
Mm_marker <- Identify_Markers(Mm_seurat,PowerCutoff=0.35, DifferenceCutoff=0.1,PvalueCutoff=0.05)
### format markers table
Zf_marker <- Format_Markers_Frac(Zf_marker)
Mm_marker <- Format_Markers_Frac(Mm_marker)
```

# 2.Identify evolutionarily conserved cell types
   CACIMAR identifies evolutionarily conserved cell types and poorly conserved cell types using the conservation score of cell types (CSCT) and a phylogenetic tree. 
  
  CSCT is calculated based on the powers (or MP values) of shared markers in each pair of cell types, relative to the sum of the powers of all markers in each pair of cell types. Shared markers indicate the markers of one species that have homologs in the other species. If genes have multiple homologs, the gene with the highest power is selected. If the pair of cell types is mutually matched, and CSCT is greater than the third quartile of CSCT values, then this pair of cell types is considered a conserved cell type between two species. Mutually matched pair cell types mean that for every cell type in species 1, we look for the corresponding cell type in species 2 that has the highest CSCT value, and vice versa. The corresponding cell types in both species are said to be “mutually matched,” and they have the highest CSCT value relative to all other cell types in the other species.
  
  Besides, we generate a phylogenetic tree based on the CSCT value using the hierarchical clustering method UPGMA (unweighted pair group method using arithmetic mean) to further identify the conserved cell types and the poorly conserved cell types. If cell types from two species are from the same clade of bipartition, the two cell types are considered conserved cell types across species. However, if these conserved cell types are only identified by the phylogenetic tree of cell types, they are regarded as poorly conserved cell types.

## (1) Inputs data of this part

- ***Marker tables for two species***
  it Should contain three column: 'gene', 'cluster' and 'power'. Marker tables can be generated by 'Identify_Markers' function in part one above. Or, you can use following codes to load test data contained in CACIMAR.

```
load(system.file("extdata", "zf_mm_markers.rda", package = "CACIMAR"))
```

## (2) Homologous gene database
  Homologous gene database is needed. Here, CACIMAR has already generated Homologous gene database for human, mouse, zebrafish and chicken, you can directly use the following codes to load database:
  
```{r}
### load homologous gene database of human and mouse
OrthG <- OrthG_Hs_Mm
### load homologous gene database of human and chicken
OrthG <- OrthG_Hs_Ch
### load homologous gene database of human and zebrafish
OrthG <- OrthG_Hs_Zf
### load homologous gene database of mouse and chicken
OrthG <- OrthG_Mm_Ch
### load homologous gene database of mouse and zebrafish
OrthG <- OrthG_Mm_Zf
### load homologous gene database of zebrafish and chicken
OrthG <- OrthG_Zf_Ch
```

  In addition, you can use function in 'Build homologous gene database' part to generate your own homologous gene database if it does not provided here.

## (3) Identify evolutionarily conserved cell types

```r
### identify evolutionarily conserved cell types based on conservation score of cell types (CSCT)
OrthG <- OrthG_Mm_Zf
expression <- Identify_ConservedCellTypes(OrthG, Zf_marker, Mm_marker,'zf','mm')
SCT_matrix <- expression[[2]]
SNT_h <- SNT[grep('mm',rownames(SCT_matrix)),as.numeric(grep('zf',colnames(SCT_matrix)))]
### show the CSCT value with a heatmap
Heatmap_Cor(SNT_h,cluster_cols=F, cluster_rows=F,Color1 = c(rgb(102/255,46/255,115/255),rgb(31/255,153/255,139/255),rgb(251/255,232/255,48/255)))
```

![cross-species cell types](CACIMAR_tutorial_files/figure-html/conserved_celltypes_heatmap.png)

## (4) Further evaluate evolutionarily conserved cell types and poorly conserved cell types through phylogenetic tree
  By default, dark blue shows the evolutionarily conserved cell types, light blue shows the poorly conserved cell types, and grey shows the unconserved cell types. You can use tiplab_cols to set the targeted colors you want. Besides, tree_method can be set to "classic Neighbor-Joining" if you want to change the method used to construct the phylogenetic tree.

```r
### get the conserved cell types based on the mutually matched and bigger than 3/4 CSCT score
conserved_hm_celltypes <- get_conserved_hm_celltypes(SNT_h)
### generate a phylogenetic tree
p <- Plot_phylogenetic_tree(SCT_matrix = SCT_matrix,
                       tree_method = "hierarchical clustering",
                       species.vector = species.vector, 
                       conserved_hm_celltype = conserved_hm_celltypes,
                       colors_labels = c("Mouse", "Zebrafish")) 
```

![cross-species cell types](CACIMAR_tutorial_files/figure-html/UPGMA_Tree.fontcolors2.png)

```r
### show the conservation of cell types in data frame
head(p$conserved_table)
```


# 3.Identify evolutionarily conserved markers
  CACIMAR first utilizes **homologous genes database** to refine markers found by "FindAllMarkers" function using ROC. And markers that exist in homologous database will be retained as shared markers. Then, CACIMAR selects markers that are in the same cell type between two species as evolution-conserved markers.

## (1) Inputs data of this part
- ***Marker tables for two species***
  First column should be gene name, second column should be cluster corresponding to marker gene. Marker tables can be generated by 'Identify_Markers' function in part one above. Or, you can use following codes to load test data.

```r
load(system.file("extdata", "zf_mm_markers.rda", package = "CACIMAR"))
```

## (2) Homologous gene database
```r
OrthG <- OrthG_Mm_Zf
```

## (3) Find evolutionarily conserved markers 
```{r message=FALSE, eval=FALSE}
ConservedMarker <- Identify_ConservedMarkers(OrthG, Mm_marker, Zf_marker,
                                               Species_name1 = 'mm', Species_name2 = 'zf')
```

### Plot conserved markers
```r
###Adjust the format to make figures of conserved markers
MarkersPlot <- FormatConservedMarkers(ConservedMarker)
###plot conserved markers for mouse
Plot_MarkersHeatmap(MarkersPlot[[1]],cellheight = 1.5,cellwidth = 6,legend = T)
```
![Mm conserved markers](CACIMAR_tutorial_files/figure-html/mm_conserved_marker2.png)
```r
###plot conserved markers for zebrafish
Plot_MarkersHeatmap(MarkersPlot[[2]],cellheight = 1.5,cellwidth = 6,legend = T)
```
![Zf conserved markers](CACIMAR_tutorial_files/figure-html/zf_conserved_marker2.png)

### Plot species-specific markers
```{r, eval=FALSE}
### Idnetiy Mm specific markers
Mm_specific_markers <- Mm_marker[!Mm_marker$gene %in% ConservedMarker$mmgene
                            ,c(-1,-3,-4,-5)]
Plot_MarkersHeatmap(Mm_specific_markers,cellheight = 0.2,cellwidth = 6)
### Identify Zf specific markers
Zf_specific_markers <- Zf_marker[!Zf_marker$gene %in% ConservedMarker$zfgene
                            ,c(-1,-3,-4,-5)]
Plot_MarkersHeatmap(Zf_specific_markers,cellheight = 0.5,cellwidth = 6)
```


# 4. Identify conserved intracellular regulation
  We utilized CACIMAR to examine the evolutionary conservation of two distinct regulatory networks: cell type-specific regulatory networks and regulatory subnetworks, also known as modules, across different species. CACIMAR calculates the conservation score of regulatory networks (CSRN) among cell types/modules in regulatory networks of different species. SCN is calculated based on two criteria: (I) fraction of homologous genes (nodes) and all genes; (II) fraction of interactions (edges) among homologous genes and all interactions. A higher CSRN indicates a more conserved regulatory network. 

## (1) Inputs data of this part
- ***Cell type specific network or Modularized regulatory networks for two species***
  Each network should contain 4 columns: 'Source', 'SourceGroup', 'Target', 'TargetGroup'. Modularized regulatory networks can be generated by [IReNA](https://github.com/jiang-junyao/IReNA) or [WGCNA](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/). Or, you can use following codes to load test data in this part.

```r
load(system.file("extdata", "gene_network.rda", package = "CACIMAR"))
```
- ***Homologous gene database***
```{r}
OrthG <- OrthG_Mm_Zf
```

## (2) Cell type specific regulatory networks analysis
  The following cell type specific regulatory networks are constructed by [IReNA](https://github.com/jiang-junyao/IReNA), which can integrate scRNA-seq and scATAC-seq/ATAC-seq data or use scRNA-seq/RNA-seq data alone to construct modularized regulatory networks. 

```{r warning=FALSE,message=FALSE, eval=FALSE}
ConservedNetworks_ct <- identify_ct_ConservedNetworks(OrthG, Species1_GRN, Species2_GRN, 'mm', 'zf')
Heatmap_Cor(ConservedNetworks_ct[[2]], cluster_cols=F, cluster_rows=F, Color1 = c(rgb(102/255,46/255,115/255), rgb(31/255,153/255,139/255), rgb(251/255,232/255,48/255)))
```

![CSRN of cell type-specific regulatory networks](CACIMAR_tutorial_files/figure-html/cell_type_network_NCS_top_regulations_300.png)

## (3) Modularized regulatory networks analysis
  Modular regulatory networks used here are constructed by R package [IReNA](https://github.com/jiang-junyao/IReNA), which can integrate scRNA-seq and scATAC-seq/ATAC-seq data or use scRNA-seq/RNA-seq data alone to construct modularized regulatory networks. [WGCNA](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/) is an optional method that can generate the input. However, WGCNA only can construct regulatory networks through bulk RNA-seq data.

```{r warning=FALSE,message=FALSE, eval=FALSE}
ConservedNetworks <- Identify_ConservedNetworks(OrthG, mm_gene_network, zf_gene_network, 'mm', 'zf')
Heatmap_Cor(ConservedNetworks[[2]], cluster_cols=F, cluster_rows=F, Color1 = c(rgb(102/255,46/255,115/255), rgb(31/255,153/255,139/255), rgb(251/255,232/255,48/255)))
```

![CSRN of modularized regulatory networks](CACIMAR_tutorial_files/figure-html/NCS_module.png)

# 5.Identify conserved intercellular interactions (cell-cell interaction)
  CACIMAR identifies evolutionarily conserved cell-cell interactions (CCI) based on the level of ligands and receptors using the SingleCellSignalR algorithm from [liana](https://saezlab.github.io/liana/). CCI that fulfill the following criteria will be considered as evolutionarily conserved: (I) the interactions between ligands and receptors are consistent; (II) the cell type of the ligand and receptor in species 1 should match the cell type of the ligand and receptor in species 2, respectively. Then, we calculate the proportion of consistent intercellular interactions to measure the conservation score for intercellular interactions (CSII). 

## (1) Inputs data of this part
- ***seurat object***
```{r, eval=FALSE}
# load(system.file("extdata", "Mm_Zf_seurat_object.Rda", package = "CACIMAR"))
```

- ***Homologous gene database***
```{r}
OrthG <- OrthG_Mm_Zf
```

## (2) perform cell-cell interaction analysis with SingleCellSignalR algorithm 
```{r, eval=FALSE}
# Mouse analysis
SingleCellSignalR_mouse_result <- perform_CCI_analysis(seurat_obj=Mm_seurat, target_organism=10090)
# Zebrafish analysis
SingleCellSignalR_zebrafish_result <- perform_CCI_analysis(seurat_obj=Zf_seurat, target_organism=7955)
# filter interactions with LRscore > 0.5
SingleCellSignalR_mouse_result2 <- subset(SingleCellSignalR_mouse_result, LRscore > 0.5)
SingleCellSignalR_zebrafish_result2 <- subset(SingleCellSignalR_zebrafish_result, LRscore > 0.5)
```

## (3) overall cell-cell interactions profile across two species
 Here, we summarize the weight of all interactions within each pair of cell types and display the profile with a sankey plot. The left side represents the source of ligands in mouse, the right side represents the source of ligands in zebrafish, and the middle represents all the target receptors shared by both species.
 
```{r, eval=FALSE}
# sum Weights of each cell types
all_weight_df_long <- calculate_Weights(species1_cci = SingleCellSignalR_mouse_result2,
                                        species2_cci = SingleCellSignalR_zebrafish_result2)
# used self-defined colors
colors_file = "/data2/jinlianli/CACIMAR/git/Tree3_3/zf_mm_colors.txt"
colors <- read.table(colors_file)
# sankey plot will be saved in current directory by default
# If the colors_file is set to NULL, the colors will be automatically assigned.
create_sankey(links = all_weight_df_long[, c("Source2", "target", "scale_weight")],
              output_file = "sankey_scale_weight.html",
              specie_name1 = "Mm",
              specie_name2 = "Zf",
              colors_file = "/data2/jinlianli/CACIMAR/git/Tree3_3/zf_mm_colors.txt"
)
```

![cell-cell interaction weight profile across species](CACIMAR_tutorial_files/figure-html/sankey_scale_weight.html_github1.png)
```{r, eval=FALSE}
# show the weight result
head(all_weight_df_long)
            source       target    weight species scale_weight             Source            Source2
1     Activated MG Activated MG 152.07967      Mm  0.008256281     MmActivated MG     MmActivated MG
2       Astrocytes Activated MG 230.97254      Mm  0.012539310       MmAstrocytes       MmAstrocytes
3            Cones Activated MG  51.89929      Mm  0.002817570            MmCones            MmCones
4     GABAergic AC Activated MG  99.92955      Mm  0.005425093     MmGABAergic AC     MmGABAergic AC
5   Glycinergic AC Activated MG  78.95619      Mm  0.004286467   MmGlycinergic AC   MmGlycinergic AC
6 Horizontal cells Activated MG 106.23687      Mm  0.005767513 MmHorizontal cells MmHorizontal cells

# show the colors dataframe which was used
head(colors)
  colors     celltype
1 #F7931E   Resting MG
2 #F8736A Activated MG
3 #A9A9A9         Rods
4 #96CEB4       Rod BC
5 #A3A500   ON cone BC
6 #C0C130  OFF cone BC
```

## (4) identified the evolutionarily conserved cell-cell interactions (CCI)
```{r, eval=FALSE}
## caculate the conservation score for intercellular interactions(CSII)

#########
######### continue this part! #########
#########



# load the test data
load(system.file("extdata", "CCC_conserved_sumary.rda", package = "CACIMAR"))

head(CCC_conserved_sumary)
# Source           Target con_weight con_num all_sp1_weight all_sp2_weight all_sp1_num all_sp2_num  score_num score_weight
# 1 Activated MG     Activated MG   9.872400      12      152.07967      12.666744         210          15 0.05333333   0.05992482
# 2 Activated MG            Cones   6.548038       8       32.33672      10.590558          44          13 0.14035088   0.15253794
# 3 Activated MG     GABAergic AC   6.693321       8       48.67964       5.077178          67           6 0.10958904   0.12451109
# 4 Activated MG   Glycinergic AC   6.536769       8       54.39586       6.566076          78           8 0.09302326   0.10722706
# 5 Activated MG Horizontal cells   8.007969      10      116.21857       7.431489         164           9 0.05780347   0.06476316
# 6 Activated MG        Microglia  13.821581      18      147.82271      18.743382         206          23 0.07860262   0.08297956

## ChordDiagram shows the CSII 
# data for ChordDiagram
cci_data <- CCC_conserved_sumary[, c("Source", "Target", "score_weight")]

# colors for ChordDiagram
# if you don't want to make colors for ChordDiagram, just set grid.col = NULL
colors <- c(rgb(247/255,147/255,30/255), rgb(248/255,115/255,106/255), rgb(169/255,169/255,169/255),
            rgb(150/255,206/255,180/255), rgb(163/255,165/255,0/255), rgb(192/255,193/255,48/255),
            rgb(157/255,115/255,194/255), rgb(183/255,76/255,171/255), rgb(230/255,134/255,201/255),
            rgb(140/255,198/255,63/255), rgb(255/255,191/255,15/255), rgb(103/255,199/255,193/255),
            rgb(3/255,161/255,198/255), rgb(97/255,156/255,255/255), rgb(129/255,70/255,58/255),
            rgb(0/255,114/255,189/255), rgb(74/255,76/255,191/255))
grid.col_name <- c("Resting MG", "Activated MG", "Rods", "Rod BC", "ON cone BC", "OFF cone BC", "GABAergic AC", "Glycinergic AC", "Cones", "RGC", "Microglia", "V/E cells", "Horizontal cells", "Astrocytes", "RPE", "Pericytes", "Reticulocytes")
names(colors) <- grid.col_name
grid.col <- colors[c(unique(cci_data$Source), unique(cci_data$Target))]

# plot ChordDiagram
ChordDiagram(net = cci_data,
             grid.col = grid.col,
             order_grid = grid.col_name,
             filename = "chordDiagram_cutoff0.75.pdf",
             link_colors_threshold = 0.75
)
```

![conservation score for intercellular interactions](CACIMAR_tutorial_files/figure-html/chordDiagram_cutoff0.75.png)

# Build homologous gene database
  Under the situation that homologous gene database of your interested species is not included in our package, CACIMAR provides to build your own homologous gene database according to MGI database. 

*Hint* Only github version of CACIMAR supports this part

## (1) Inputs data of this part

- ***MGI database***
  MGI database can be downloaded from http://www.informatics.jax.org/downloads/reports/HOM_AllOrganism.rpt.

- ***species name***
  Currently CACIMAR is able to build homologous gene database for the following species:
  
```
1) 'mm' for mouse; 
2) 'zf' for zebrafish; 
3) 'hs' for human; 
4) 'ch' for chicken; 
5) 'cf' for dog; 
6) 'pt' for chimpanzee; 
7) 'xt' for frog; 
8) 'rn' for rat; 
9) 'bt' for cattle; 
10) 'rh' for macaque
```

## (2) Example for build homologous gene database
```r
### download data
download.file(destfile = 'D:\\GIBH\\platform\\OrthG database/HOM.txt',url = 'http://www.informatics.jax.org/downloads/reports/HOM_AllOrganism.rpt')
### load data
HOM=read.delim('D:\\GIBH\\platform\\OrthG database/HOM.txt')
### build homologous gene database for dog and mouse
Species_name1 <- 'cf'
Species_name2 <- 'mm'
OrthG_Cf_Mm <- buildHomDatabase(HOM, Species_name1, Species_name2)
```



